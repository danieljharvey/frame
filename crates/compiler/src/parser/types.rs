use std::cell::RefCell;
use std::ops::Range;

use crate::types::{Prim, TypePrim};

pub type IResult<'a, T> = nom::IResult<LocatedSpan<'a>, T>;

pub type LocatedSpan<'a> = nom_locate::LocatedSpan<&'a str, State<'a>>;

#[derive(Debug, Clone, PartialEq)]
pub struct Annotation<Span> {
    pub start: Span,
    pub end: Span,
}

pub type ParseAnnotation<'a> = Annotation<LocatedSpan<'a>>;

pub type StaticAnnotation<'a> = Annotation<nom_locate::LocatedSpan<&'a str, ()>>;

pub trait ToRange {
    fn to_range(&self) -> Range<usize>;
}

// this suggests we're using LocatedSpan wrong
impl<'a> ToRange for LocatedSpan<'a> {
    fn to_range(&self) -> Range<usize> {
        let start = self.location_offset();
        let end = start + self.fragment().len();
        start..end
    }
}

#[derive(Debug, Clone)]
pub struct ParseError {
    pub range: Range<usize>,
    pub message: String,
}

#[derive(Clone, Debug)]
pub struct State<'a>(pub &'a RefCell<Vec<ParseError>>);

impl<'a> State<'a> {
    pub fn report_error(&self, error: ParseError) {
        self.0.borrow_mut().push(error);
    }
}

#[derive(Debug)]
pub enum ParseExpr<'a> {
    Ident {
        ann: ParseAnnotation<'a>,
        var: String,
    },
    Prim {
        ann: ParseAnnotation<'a>,
        prim: Prim,
    },
    If {
        ann: ParseAnnotation<'a>,
        pred_expr: Box<ParseExpr<'a>>,
        then_expr: Box<ParseExpr<'a>>,
        else_expr: Box<ParseExpr<'a>>,
    },
    Ann {
        ann: ParseAnnotation<'a>,
        ty: Option<ParseType<'a>>,
        expr: Box<ParseExpr<'a>>,
    },
    Error,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParseType<'a> {
    Prim {
        ann: ParseAnnotation<'a>,
        type_prim: TypePrim,
    },
}
