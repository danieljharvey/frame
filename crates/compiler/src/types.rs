use std::fmt::{Display, Error, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub enum Prim {
    Boolean(bool),
    IntLit(i64),
}

impl Display for &Prim {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            Prim::Boolean(true) => write!(f, "True"),
            Prim::Boolean(false) => write!(f, "False"),
            Prim::IntLit(i) => i.fmt(f),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum TypePrim {
    TBoolean,
    TInt8,
    TInt16,
    TInt32,
    TInt64,
}

impl Display for &TypePrim {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            TypePrim::TBoolean => write!(f, "Boolean"),
            TypePrim::TInt8 => write!(f, "Int8"),
            TypePrim::TInt16 => write!(f, "Int16"),
            TypePrim::TInt32 => write!(f, "Int32"),
            TypePrim::TInt64 => write!(f, "Int64"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Type<Ann> {
    TPrim { ann: Ann, type_prim: TypePrim },
}

impl<Ann> Display for &Type<Ann> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            Type::TPrim { type_prim, .. } => type_prim.fmt(f),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr<Ann> {
    EPrim {
        ann: Ann,
        prim: Prim,
    },
    EIf {
        ann: Ann,
        pred_expr: Box<Expr<Ann>>,
        then_expr: Box<Expr<Ann>>,
        else_expr: Box<Expr<Ann>>,
    },
    EAnn {
        ann: Ann,
        ty: Type<Ann>,
        expr: Box<Expr<Ann>>,
    },
}

impl<Ann> Display for Expr<Ann> {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        match self {
            Expr::EPrim { prim, .. } => prim.fmt(f),
            Expr::EIf {
                pred_expr,
                then_expr,
                else_expr,
                ..
            } => write!(
                f,
                "if {} then {} else {}",
                &pred_expr, &then_expr, &else_expr
            ),
            Expr::EAnn { ty, expr, .. } => write!(f, "({}: {})", &expr, ty),
        }
    }
}
