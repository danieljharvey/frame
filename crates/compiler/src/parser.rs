pub mod constructors;

use std::cell::RefCell;
use std::ops::Range;

use crate::types::{Expr, Prim, Type, TypePrim};
use nom::branch::alt;
use nom::bytes::complete::{take, take_till1, take_while};
use nom::character::complete::{anychar, multispace0};
use nom::combinator::{all_consuming, map, not, recognize, rest, verify};
use nom::sequence::{preceded, terminated};
use nom::{
    bytes::complete::{tag, take_while_m_n},
    combinator::map_res,
};

type IResult<'a, T> = nom::IResult<LocatedSpan<'a>, T>;

pub type LocatedSpan<'a> = nom_locate::LocatedSpan<&'a str, State<'a>>;

#[derive(Debug, Clone, PartialEq)]
pub struct Annotation<Span> {
    pub start: Span,
    pub end: Span,
}

type ParseAnnotation<'a> = Annotation<LocatedSpan<'a>>;

pub type StaticAnnotation<'a> = Annotation<nom_locate::LocatedSpan<&'a str, ()>>;

trait ToRange {
    fn to_range(&self) -> Range<usize>;
}

// this suggests we're using LocatedSpan wrong
impl<'a> ToRange for LocatedSpan<'a> {
    fn to_range(&self) -> Range<usize> {
        let start = self.location_offset();
        let end = start + self.fragment().len();
        start..end
    }
}

#[derive(Debug, Clone)]
pub struct Error<Span> {
    pub annotation: Annotation<Span>,
    pub message: String,
}

pub type ParseError<'a> = Error<LocatedSpan<'a>>;

pub type StaticParseError<'a> = Error<nom_locate::LocatedSpan<&'a str, ()>>;

#[derive(Clone, Debug)]
pub struct State<'a>(&'a RefCell<Vec<ParseError<'a>>>);

impl<'a> State<'a> {
    pub fn report_error(&self, error: ParseError<'a>) {
        self.0.borrow_mut().push(error);
    }
}

// given a parser, return a parser that returns the result and an Annotation describing the source
// location
pub fn with_annotation<'a, F, T>(
    mut inner: F,
) -> impl FnMut(LocatedSpan<'a>) -> IResult<(ParseAnnotation<'a>, T)>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
{
    move |input: LocatedSpan| {
        let (input, start) = nom_locate::position(input)?;
        let (input, result) = inner(input)?;
        let (input, end) = nom_locate::position(input)?;
        Ok((input, (Annotation { start, end }, result)))
    }
}

fn expect<'a, F, E, T>(
    mut parser: F,
    error_msg: E,
) -> impl FnMut(LocatedSpan<'a>) -> IResult<Option<T>>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
    E: ToString,
{
    move |input| match parser(input) {
        Ok((remaining, out)) => Ok((remaining, Some(out))),
        Err(nom::Err::Error(nom::error::Error { input, .. }))
        | Err(nom::Err::Failure(nom::error::Error { input, .. })) => {
            let err = Error {
                annotation: Annotation {
                    start: input.clone(),
                    end: input.clone(),
                },
                message: error_msg.to_string(),
            };
            input.extra.report_error(err);
            Ok((input, None))
        }
        Err(err) => Err(err),
    }
}

#[derive(Debug)]
pub struct Ident(String);

#[derive(Debug)]
pub enum ParseExpr<'a> {
    Ident {
        ann: ParseAnnotation<'a>,
        ident: Ident,
    },
    Prim {
        ann: ParseAnnotation<'a>,
        prim: Prim,
    },
    If {
        ann: ParseAnnotation<'a>,
        pred_expr: Box<ParseExpr<'a>>,
        then_expr: Box<ParseExpr<'a>>,
        else_expr: Box<ParseExpr<'a>>,
    },
    Ann {
        ann: ParseAnnotation<'a>,
        ty: Option<ParseType<'a>>,
        expr: Box<ParseExpr<'a>>,
    },
    Error,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ParseType<'a> {
    Prim {
        ann: ParseAnnotation<'a>,
        type_prim: TypePrim,
    },
}

fn ident(input: LocatedSpan) -> IResult<ParseExpr> {
    let first = verify(anychar, |c| c.is_ascii_alphabetic() || *c == '_');
    let rest = take_while(|c: char| c.is_ascii_alphanumeric() || "_-'".contains(c));
    let ident = ws(with_annotation(recognize(preceded(first, rest))));
    map(ident, |(ann, span): (ParseAnnotation, LocatedSpan)| {
        ParseExpr::Ident {
            ann,
            ident: Ident(span.fragment().to_string()),
        }
    })(input)
}

// Need to add more types as we need them pls
fn parse_type_prim(input: LocatedSpan) -> IResult<TypePrim> {
    let boolean = map(tag("Boolean"), |_| TypePrim::TBoolean);
    let int_8 = map(tag("Int8"), |_| TypePrim::TInt8);
    let int_16 = map(tag("Int16 "), |_| TypePrim::TInt16);
    let int_32 = map(tag("Int32"), |_| TypePrim::TInt32);
    let int_64 = map(tag("Int64"), |_| TypePrim::TInt64);
    ws(alt((boolean, int_8, int_16, int_32, int_64)))(input)
}

pub fn ty(input: LocatedSpan) -> IResult<ParseType> {
    let (input, (ann, type_prim)) = ws(with_annotation(parse_type_prim))(input)?;
    Ok((input, ParseType::Prim { ann, type_prim }))
}

fn iff(input: LocatedSpan) -> IResult<ParseExpr> {
    let (
        input,
        (
            ann,
            ParsedIf {
                pred_expr,
                then_expr,
                else_expr,
            },
        ),
    ) = ws(with_annotation(parse_if_internal))(input)?;

    Ok((
        input,
        ParseExpr::If {
            ann,
            pred_expr: Box::new(pred_expr),
            then_expr: Box::new(then_expr),
            else_expr: Box::new(else_expr),
        },
    ))
}

struct ParsedIf<'a> {
    pred_expr: ParseExpr<'a>,
    then_expr: ParseExpr<'a>,
    else_expr: ParseExpr<'a>,
}

fn parse_if_internal(input: LocatedSpan) -> IResult<ParsedIf> {
    let (input, _) = tag("if")(input)?;
    let (input, pred_expr) = expect(expr, "expected expression after 'if'")(input)?;

    let (input, _) = ws(tag("then"))(input)?;
    let (input, then_expr) = expect(expr, "expected expression after 'then'")(input)?;

    let (input, _) = ws(tag("else"))(input)?;
    let (input, else_expr) = expect(expr, "expected expression after 'else'")(input)?;

    Ok((
        input,
        ParsedIf {
            pred_expr: pred_expr.unwrap_or(ParseExpr::Error),
            then_expr: then_expr.unwrap_or(ParseExpr::Error),
            else_expr: else_expr.unwrap_or(ParseExpr::Error),
        },
    ))
}

fn error(input: LocatedSpan) -> IResult<ParseExpr> {
    map(take_till1(|c| c == ')'), |span: LocatedSpan| {
        let err = Error {
            annotation: Annotation {
                start: span.clone(),
                end: span.clone(),
            },
            message: format!("unexpected `{}`", span.fragment()),
        };
        span.extra.report_error(err);
        ParseExpr::Error
    })(input)
}

// our main Expr parser, basically, try all the parsers
fn expr(input: LocatedSpan) -> IResult<ParseExpr> {
    alt((ann, prim, iff, ident, error))(input)
}

// parse a whole source file
fn source_file(input: LocatedSpan) -> IResult<ParseExpr> {
    let expr = alt((expr, map(take(0usize), |_| ParseExpr::Error)));
    terminated(expr, preceded(expect(not(anychar), "expected EOF"), rest))(input)
}

pub fn parse<'a, 'state>(
    errors: &'state RefCell<Vec<ParseError<'a>>>,
    source: &'a str,
) -> (ParseExpr<'a>, Vec<ParseError<'a>>)
where
    'state: 'a,
{
    let input = LocatedSpan::new_extra(source, State(&errors));
    let (_, expr) = all_consuming(source_file)(input).expect("parser cannot fail");
    let copied_errors = errors.clone().into_inner().clone();

    (expr, copied_errors)
}

pub fn parse_type<'a, 'state>(
    errors: &'state RefCell<Vec<ParseError<'a>>>,
    source: &'a str,
) -> (ParseType<'a>, Vec<ParseError<'a>>)
where
    'state: 'a,
{
    let input = LocatedSpan::new_extra(source, State(errors));
    let (_, ty) = all_consuming(ty)(input).expect("parser cannot fail");
    (ty, errors.clone().into_inner())
}

pub fn main() {
    for input in &["foo", "(foo)", "(foo))", "(%", "(", "%", "()", ""] {
        let errors = RefCell::new(Vec::new());
        println!("{:7} {:?}", input, parse(&errors, input));
    }
}

//////// prims

fn is_int_digit(c: char) -> bool {
    c.is_ascii_digit()
}

fn int_primary(input: LocatedSpan) -> IResult<u8> {
    map_res(take_while_m_n(1, 12, is_int_digit), from_int)(input)
}

fn from_int(input: LocatedSpan) -> Result<u8, std::num::ParseIntError> {
    input.parse::<u8>()
}

fn int(input: LocatedSpan) -> IResult<Prim> {
    let (input, int_val) = int_primary(input)?;
    Ok((input, Prim::IntLit(i64::from(int_val))))
}

fn parse_true(input: LocatedSpan) -> IResult<Prim> {
    map(tag("True"), |_| Prim::Boolean(true))(input)
}

fn parse_false(input: LocatedSpan) -> IResult<Prim> {
    map(tag("False"), |_| Prim::Boolean(false))(input)
}

fn bool(input: LocatedSpan) -> IResult<Prim> {
    (alt((parse_true, parse_false)))(input)
}

fn prim(input: LocatedSpan) -> IResult<ParseExpr> {
    map(ws(with_annotation(alt((bool, int)))), |(ann, prim)| {
        ParseExpr::Prim { ann, prim }
    })(input)
}

/// A combinator that takes a parser `inner` and produces a parser that also consumes both leading and
/// trailing whitespace, returning the output of `inner`.
pub fn ws<'a, F, T>(inner: F) -> impl FnMut(LocatedSpan<'a>) -> IResult<T>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
{
    preceded(multispace0, inner)
}

fn parse_ann_internal(input: LocatedSpan) -> IResult<(Option<ParseType>, Option<ParseExpr>)> {
    let (input, _) = tag("(")(input)?;

    let (input, expr) = expect(expr, "expected expression after '('")(input)?;

    let (input, _) = expect(ws(tag(":")), "expected ':'")(input)?;

    let (input, ty) = expect(ty, "expected type after ':'")(input)?;

    let (input, _) = expect(tag(")"), "expected closing ')'")(input)?;

    Ok((input, (ty, expr)))
}

fn ann(input: LocatedSpan) -> IResult<ParseExpr> {
    let (input, (ann, (ty, expr))) = ws(with_annotation(parse_ann_internal))(input)?;

    Ok((
        input,
        ParseExpr::Ann {
            ann,
            ty,
            expr: Box::new(expr.unwrap_or(ParseExpr::Error)),
        },
    ))
}

#[derive(Debug, PartialEq)]
pub enum ParseConvertError {
    FoundError,
    FoundIdent,
    MissingTypeAnnotation,
}

// turn this into our 'real' expr that does not include `Error`
// so we can typecheck it and generally try it out
pub fn to_real_expr(parse_expr: ParseExpr) -> Result<Expr<StaticAnnotation>, ParseConvertError> {
    match parse_expr {
        ParseExpr::Ann { ann, ty, expr } => Ok(Expr::EAnn {
            ann: to_real_ann(ann),
            ty: to_real_ty(ty)?,
            expr: Box::new(to_real_expr(*expr)?),
        }),
        ParseExpr::Prim { ann, prim } => Ok(Expr::EPrim {
            ann: to_real_ann(ann),
            prim,
        }),
        ParseExpr::Ident { .. } => Err(ParseConvertError::FoundIdent),
        ParseExpr::If {
            ann,
            pred_expr,
            then_expr,
            else_expr,
        } => Ok(Expr::EIf {
            ann: to_real_ann(ann),
            pred_expr: Box::new(to_real_expr(*pred_expr)?),
            then_expr: Box::new(to_real_expr(*then_expr)?),
            else_expr: Box::new(to_real_expr(*else_expr)?),
        }),
        ParseExpr::Error => Err(ParseConvertError::FoundError),
    }
}

pub fn to_real_ty(
    parse_ty: Option<ParseType>,
) -> Result<Type<StaticAnnotation>, ParseConvertError> {
    parse_ty
        .map(|ty| match ty {
            ParseType::Prim { ann, type_prim } => Type::TPrim {
                ann: to_real_ann(ann),
                type_prim,
            },
        })
        .ok_or(ParseConvertError::MissingTypeAnnotation)
}

fn to_real_ann(ann: ParseAnnotation) -> StaticAnnotation {
    Annotation {
        start: ann.start.map_extra(|_| ()),
        end: ann.end.map_extra(|_| ()),
    }
}

pub fn to_real_error(error: ParseError) -> StaticParseError {
    Error {
        annotation: to_real_ann(error.annotation),
        message: error.message,
    }
}
