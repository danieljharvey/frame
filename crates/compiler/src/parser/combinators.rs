pub use super::types::{Annotation, IResult, LocatedSpan, ParseError, Position, ToAnnotation};

// given a parser, return a parser that returns the result and an Annotation describing the source
// location
pub fn with_annotation<'a, F, T>(
    mut inner: F,
) -> impl FnMut(LocatedSpan<'a>) -> IResult<(Annotation, T)>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
{
    move |input: LocatedSpan| {
        let (input, start) = nom_locate::position(input)?;
        let (input, result) = inner(input)?;
        let (input, end) = nom_locate::position(input)?;

        let ann = Annotation {
            start: Position {
                offset: start.location_offset(),
                row: start.location_line(),
                column: start.get_utf8_column(),
            },
            end: Position {
                offset: end.location_offset(),
                row: end.location_line(),
                column: end.get_utf8_column(),
            },
        };
        Ok((input, (ann, result)))
    }
}

pub fn expect<'a, F, E, T>(
    mut parser: F,
    error_msg: E,
) -> impl FnMut(LocatedSpan<'a>) -> IResult<Option<T>>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
    E: ToString,
{
    move |input| match parser(input) {
        Ok((remaining, out)) => Ok((remaining, Some(out))),
        Err(nom::Err::Error(nom::error::Error { input, .. }))
        | Err(nom::Err::Failure(nom::error::Error { input, .. })) => {
            let err = ParseError {
                ann: input.to_annotation(),
                message: error_msg.to_string(),
            };
            input.extra.report_error(err);
            Ok((input, None))
        }
        Err(err) => Err(err),
    }
}

/// A combinator that takes a parser `inner` and produces a parser that also consumes both leading and
/// trailing whitespace, returning the output of `inner`.
pub fn ws<'a, F, T>(inner: F) -> impl FnMut(LocatedSpan<'a>) -> IResult<T>
where
    F: FnMut(LocatedSpan<'a>) -> IResult<T>,
{
    nom::sequence::preceded(nom::character::complete::multispace0, inner)
}
