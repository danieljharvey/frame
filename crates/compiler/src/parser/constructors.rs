use crate::types::{Expr, Prim};

// construct int
pub fn int<Ann>(ann: Ann, int_val: i64) -> Expr<Ann> {
    Expr::EPrim {
        ann,
        prim: Prim::IntLit(int_val),
    }
}

// construct bool
pub fn bool<Ann>(ann: Ann, bool_val: bool) -> Expr<Ann> {
    Expr::EPrim {
        ann,
        prim: Prim::Boolean(bool_val),
    }
}

// construct if
pub fn mk_if<Ann>(
    ann: Ann,
    pred_expr: Expr<Ann>,
    then_expr: Expr<Ann>,
    else_expr: Expr<Ann>,
) -> Expr<Ann> {
    Expr::EIf {
        ann,
        pred_expr: Box::new(pred_expr),
        then_expr: Box::new(then_expr),
        else_expr: Box::new(else_expr),
    }
}
