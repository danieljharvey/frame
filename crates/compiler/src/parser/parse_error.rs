use super::types::ParseError;
use std::ops::Range;

pub fn to_report(parse_error: &ParseError) -> ariadne::Report<Range<usize>> {
    use ariadne::{Label, Report, ReportKind};

    Report::build(ReportKind::Error, (), 12)
        .with_code(1)
        .with_message("Parse error")
        .with_label(
            Label::new(parse_error.ann.start.offset..parse_error.ann.end.offset)
                .with_message(parse_error.message.clone()),
        )
        .finish()
}
