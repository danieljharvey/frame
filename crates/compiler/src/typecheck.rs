mod warning;
use crate::types::{Expr, Prim, Type, TypePrim};
pub use warning::Warning;
mod type_error;
pub use type_error::{to_report, TypeError};
mod expr_utils;
pub use expr_utils::{
    get_outer_expr_annotation, get_outer_type_annotation, set_outer_expr_annotation,
    set_outer_type_annotation,
};

fn infer_prim<Ann>(ann: Ann, prim: &Prim) -> Result<Type<Ann>, TypeError<Ann>> {
    match prim {
        Prim::Boolean(_) => Ok(Type::TPrim {
            ann,
            type_prim: TypePrim::TBoolean,
        }),
        Prim::IntLit(_) => Err(TypeError::UnknownIntegerLiteral { ann }),
    }
}

// given no information, try and work out what we have here
pub fn infer<Ann: Clone>(
    expr: &Expr<Ann>,
    warnings: &mut Vec<Warning<Ann>>,
) -> Result<Expr<Type<Ann>>, TypeError<Ann>> {
    match expr {
        Expr::EPrim { ann, prim } => {
            let ty = infer_prim(ann.clone(), prim)?;
            Ok(Expr::EPrim {
                ann: ty,
                prim: prim.clone(),
            })
        }
        Expr::EIdent { ann, var } => Err(TypeError::UnknownVariable {
            var: var.clone(),
            ann: ann.clone(),
        }),
        Expr::EIf {
            ann,
            pred_expr,
            then_expr,
            else_expr,
        } => check_if(None, ann, pred_expr, then_expr, else_expr, warnings),
        Expr::EAnn { ann, ty, expr } => {
            let typed_expr = check(ty, expr, warnings)?;
            let inferred_type = get_outer_expr_annotation(&typed_expr);
            if let Type::TPrim {
                type_prim: TypePrim::TBoolean,
                ..
            } = inferred_type
            {
                // ie, we don't need this annotation
                warnings.push(Warning::UnnecessaryAnnotation { ann: ann.clone() })
            }
            let ty = set_outer_type_annotation(inferred_type, ann);
            Ok(set_outer_expr_annotation(&typed_expr, &ty))
        }
    }
}

fn check_if<Ann: Clone>(
    maybe_ty: Option<&Type<Ann>>,
    ann: &Ann,
    pred_expr: &Expr<Ann>,
    then_expr: &Expr<Ann>,
    else_expr: &Expr<Ann>,
    warnings: &mut Vec<Warning<Ann>>,
) -> Result<Expr<Type<Ann>>, TypeError<Ann>> {
    let typed_pred_expr = check(
        &Type::TPrim {
            ann: get_outer_expr_annotation(pred_expr).clone(),
            type_prim: TypePrim::TBoolean,
        },
        pred_expr,
        warnings,
    )?;
    let typed_then_expr = match maybe_ty {
        Some(ty) => check(ty, then_expr, warnings),
        None => infer(then_expr, warnings),
    }?;
    let ty_then = get_outer_expr_annotation(&typed_then_expr);
    let typed_else_expr = check(ty_then, else_expr, warnings)?;
    let ty = set_outer_type_annotation(ty_then, ann);
    Ok(Expr::EIf {
        ann: ty,
        pred_expr: Box::new(typed_pred_expr),
        then_expr: Box::new(typed_then_expr),
        else_expr: Box::new(typed_else_expr),
    })
}

// given a type, try and work out what we have here
fn check<Ann: Clone>(
    ty: &Type<Ann>,
    expr: &Expr<Ann>,
    warnings: &mut Vec<Warning<Ann>>,
) -> Result<Expr<Type<Ann>>, TypeError<Ann>> {
    match (ty, expr) {
        (
            ty,
            Expr::EIf {
                ann,
                pred_expr,
                then_expr,
                else_expr,
            },
        ) => check_if(Some(ty), ann, pred_expr, then_expr, else_expr, warnings),
        (ty, Expr::EIdent { ann, var }) => {
            // whatever the type sig says goes
            let ty_ident = set_outer_type_annotation(ty, ann);
            Ok(Expr::EIdent {
                ann: ty_ident,
                var: var.clone(),
            })
        }
        (Type::TPrim { type_prim, .. }, Expr::EPrim { prim, ann }) => {
            check_prim(ann, type_prim, prim)?;
            Ok(Expr::EPrim {
                ann: Type::TPrim {
                    ann: ann.clone(),
                    type_prim: type_prim.clone(),
                },
                prim: prim.clone(),
            })
        }
        _ => {
            let typed_expr = infer(expr, warnings)?;
            let ty_expr = get_outer_expr_annotation(&typed_expr);
            let _resolved_ty = unify(ty, ty_expr)?;
            // once we have polymorphism, we'll need to insert `resolved_ty` into `typed_expr`
            // but for now we can just return it
            Ok(typed_expr)
        }
    }
}

fn unify<Ann: Clone>(
    expected: &Type<Ann>,
    actual: &Type<Ann>,
) -> Result<Type<Ann>, TypeError<Ann>> {
    match (expected, actual) {
        (
            Type::TPrim {
                type_prim: type_prim_left,
                ..
            },
            Type::TPrim {
                type_prim: type_prim_right,
                ..
            },
        ) => {
            if type_prim_left == type_prim_right {
                Ok(expected.clone())
            } else {
                Err(TypeError::TypeMismatch {
                    expected: expected.clone(),
                    actual: actual.clone(),
                })
            }
        }
    }
}

fn check_prim<Ann: Clone>(
    ann: &Ann,
    type_prim: &TypePrim,
    prim: &Prim,
) -> Result<(), TypeError<Ann>> {
    match (type_prim, prim) {
        (TypePrim::TBoolean, Prim::Boolean(_)) => Ok(()),
        (TypePrim::TInt8, Prim::IntLit(_)) => Ok(()),
        (TypePrim::TInt16, Prim::IntLit(_)) => Ok(()),
        (TypePrim::TInt32, Prim::IntLit(_)) => Ok(()),
        (TypePrim::TInt64, Prim::IntLit(_)) => Ok(()),

        _ => Err(TypeError::LiteralMismatch {
            ann: ann.clone(),
            type_prim: type_prim.clone(),
            prim: prim.clone(),
        }),
    }
}
