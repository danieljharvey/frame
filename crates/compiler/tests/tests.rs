use frame::parser::constructors::{bool, int, mk_if};
use frame::typecheck::TypeError;
use frame::types::{Expr, Prim, Type, TypePrim};
use std::cell::RefCell;

fn void_expr<Ann>(expr: Expr<Ann>) -> Expr<()> {
    match expr {
        Expr::EPrim { prim, .. } => Expr::EPrim { ann: (), prim },
        Expr::EIf {
            pred_expr,
            then_expr,
            else_expr,
            ..
        } => Expr::EIf {
            ann: (),
            pred_expr: Box::new(void_expr(*pred_expr)),
            then_expr: Box::new(void_expr(*then_expr)),
            else_expr: Box::new(void_expr(*else_expr)),
        },
        Expr::EAnn { ty, expr, .. } => Expr::EAnn {
            ann: (),
            ty: void_type(ty),
            expr: Box::new(void_expr(*expr)),
        },
    }
}

fn void_type<Ann>(ty: Type<Ann>) -> Type<()> {
    match ty {
        Type::TPrim { type_prim, .. } => Type::TPrim { ann: (), type_prim },
    }
}
fn void_type_error<Ann>(type_error: TypeError<Ann>) -> TypeError<()> {
    match type_error {
        TypeError::TypeMismatch { expected, actual } => TypeError::TypeMismatch {
            expected: void_type(expected),
            actual: void_type(actual),
        },
        TypeError::LiteralMismatch {
            prim, type_prim, ..
        } => TypeError::LiteralMismatch {
            ann: (),
            prim,
            type_prim,
        },
        TypeError::UnknownIntegerLiteral { .. } => TypeError::UnknownIntegerLiteral { ann: () },
    }
}

#[test]
fn test_parse() {
    let tests = vec![
        (" 1", int((), 1)),
        ("1", int((), 1)),
        ("11", int((), 11)),
        ("    11dog", int((), 11)),
        (" True", bool((), true)),
        ("False", bool((), false)),
        ("    True100", bool((), true)),
        (
            "if 1 then False else True",
            mk_if((), int((), 1), bool((), false), bool((), true)),
        ),
        (
            "if True then 1 else 2",
            mk_if((), bool((), true), int((), 1), int((), 2)),
        ),
        (
            "(1: Int64)",
            Expr::EAnn {
                ann: (),
                ty: Type::TPrim {
                    ann: (),
                    type_prim: TypePrim::TInt64,
                },
                expr: Box::new(int((), 1)),
            },
        ),
    ];

    for (input, expect) in tests {
        let ref_cell = RefCell::new(vec![]);
        let (parse_result, _) = frame::parser::parse(&ref_cell, input);
        let result = frame::parser::to_real_expr(parse_result);
        dbg!(&result);
        let voided_result = result.map(void_expr);
        assert_eq!(voided_result, Ok(expect))
    }
}

#[test]
fn test_typecheck_success() {
    let tests = vec![
        ("True", "Boolean"),
        ("False", "Boolean"),
        ("if True then False else True", "Boolean"),
        ("if True then (1: Int64) else 2", "Int64"),
        //        ("(if True then 1 else 2 : Int64)", "Int64"),
    ];

    for (input, expected) in tests {
        let ref_cell = RefCell::new(vec![]);
        let (parse_result, _) = frame::parser::parse(&ref_cell, input);
        let input_expr = frame::parser::to_real_expr(parse_result).expect("parsing expr");

        let (expected_parse_type, _) = frame::parser::parse_type(&ref_cell, expected);
        let expected_type =
            frame::parser::to_real_ty(Some(expected_parse_type)).expect("parsing type");

        let typed_expr = frame::typecheck::infer(&void_expr(input_expr), &mut vec![])
            .expect("should have succeeded");
        let ty = frame::typecheck::get_outer_expr_annotation(&typed_expr);

        assert_eq!(ty, &void_type(expected_type));
    }
}

#[test]
fn test_typecheck_failure() {
    let tests = vec![
        (
            "if True then 1 else 2",
            TypeError::UnknownIntegerLiteral { ann: () },
        ),
        (
            "if (1: Int64) then True else False",
            TypeError::TypeMismatch {
                expected: Type::TPrim {
                    ann: (),
                    type_prim: TypePrim::TBoolean,
                },
                actual: Type::TPrim {
                    ann: (),
                    type_prim: TypePrim::TInt64,
                },
            },
        ),
        (
            "if True then True else (1: Int64)",
            TypeError::TypeMismatch {
                expected: Type::TPrim {
                    ann: (),
                    type_prim: TypePrim::TBoolean,
                },
                actual: Type::TPrim {
                    ann: (),
                    type_prim: TypePrim::TInt64,
                },
            },
        ),
        (
            "(1: Boolean)",
            TypeError::LiteralMismatch {
                ann: (),
                type_prim: TypePrim::TBoolean,
                prim: Prim::IntLit(1),
            },
        ),
    ];

    for (input, expected_type_error) in tests {
        let ref_cell = RefCell::new(vec![]);
        let (parse_result, _) = frame::parser::parse(&ref_cell, input);
        let input_expr = frame::parser::to_real_expr(parse_result).expect("parsing expr");

        let result = frame::typecheck::infer(&input_expr, &mut vec![]);

        let type_error = match result {
            Err(e) => e,
            Ok(_) => panic!("should not get here"),
        };

        assert_eq!(void_type_error(type_error.clone()), expected_type_error);

        let mut buf = std::io::BufWriter::new(Vec::new());

        let report = frame::typecheck::to_report(&type_error);

        report
            .write(ariadne::Source::from(input), &mut buf)
            .unwrap();

        let bytes = buf.into_inner().unwrap();
        let string = String::from_utf8(bytes).unwrap();

        println!("{string}");
        insta::assert_snapshot!(string);
    }
}
