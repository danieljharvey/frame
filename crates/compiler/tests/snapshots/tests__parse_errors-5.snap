---
source: crates/compiler/tests/tests.rs
expression: parse_result
---
ParseBlock {
    let_bindings: [],
    final_expr: Ident {
        ann: Annotation {
            start: Position {
                offset: 0,
                row: 1,
                column: 1,
            },
            end: Position {
                offset: 3,
                row: 1,
                column: 4,
            },
        },
        var: "dog",
    },
}
